### ISSUE TITLE
[❔ other question] How to get the derivatives of 2d samples generated from an independent sampler

### ISSUE THREAD

**Issue by xchhuang** [mitsuba-renderer/mitsuba2#444]
## Summary

*Is there an easy way to get the derivatives of 2d samples(e.g., used for emitter sampling) generated from a sampler?*

## Description
For now we can get the derivative of the parameters(e.g., reflectance) within a `scene` using the differentiable `render` function:
```
image = render(scene, optimizer=opt, unbiased=True, spp=1)
```
Can we also get the derivative of 2d samples for some optimization purpose from the following independent sampler? Thanks in advance.
```
<sampler type="independent">  <!-- ldsampler -->
    <integer name="sample_count" value="$spp"/>
</sampler>
```

**Comment by Speierers** [comment]
Hi @xchhuang ,

I believe this should be possible although this wouldn't work with the `traverse` mechanism. You will have to do things manually.

For instance, in your integrator (e.g. defined in Python), you could draw a 2D sample and then ask the AD system to track computations using it like this:
```python
samples = sampler.next_float2()
ek.set_requires_gradient(samples, True)

# some fancy computations
loss = foo(samples) 
ek.backward(loss)
print(ek.gradient(samples))
```

Please let me know if this answers your question?

**Comment by xchhuang** [comment]
Hi @Speierers ,

Thanks, that makes sense. Just have one more question: if we follow the `Integrator Example` in https://mitsuba2.readthedocs.io/en/latest/src/advanced_topics/custom_plugins.html:

After calling for example:
```
samples = sampler.next_float2()
ek.set_requires_gradient(samples, True)
rays = GetRayFromSamples(samples)
scene.ray_intersect(rays)
...
scene.sample_emitter_direction
...
register_integrator("mydirectintegrator", lambda props: MyDirectIntegrator(props))
...
scene.integrator().render(scene, scene.sensors()[0])
```
is there anyway to define the loss based on the rendered image? Or we need to re-write the pipeline ourselves with some differentiable operators? Thanks in advance.

**Comment by Speierers** [comment]
Hi @xchhuang,

I don't think you will be able to access the data on the `film` in a differentiable way. Please take a look at `/src/python/python/autodiff.py:render_helper()` to see how to write a custom differentiable pipeline yourself.

**Comment by xchhuang** [comment]
> Hi @xchhuang,
> 
> I don't think you will be able to access the data on the `film` in a differentiable way. Please take a look at `/src/python/python/autodiff.py:render_helper()` to see how to write a custom differentiable pipeline yourself.

I see. Thank you I will take a look.
